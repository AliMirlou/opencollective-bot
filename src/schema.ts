import * as joi from 'joi'

export interface Config {
  labels?: string[]
  message: string
  opencollective: string
}

/* Schema, Defaults */

export const defaultMessage = `
Hey :wave:,

We saw you use our package. To keep our code as accessible as possible, we decided to open source it and are now helping others with their questions.

In a pursuit to continue our work, help us by donating to our collective! :heart:
`

export const schema = joi.object().keys({
  labels: joi
    .array()
    .single()
    .description(
      'Issues to apply message to. By default, every issue will receive the message.',
    )
    .default(false),
  message: joi
    .string()
    .required()
    .description('Message for your backers.')
    .default(defaultMessage),
  opencollective: joi
    .string()
    .regex(new RegExp('^https://opencollective.com/.+'))
    .required()
    .description('Open collective link of your collective.'),
})

export async function validateSchema(config: Config): Promise<Config | null> {
  try {
    const value = await joi.validate(config, schema)

    return value
  } catch (err) {
    return null
  }
}
