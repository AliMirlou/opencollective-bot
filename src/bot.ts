import * as probot from 'probot'
import getConfig from 'probot-config'
import { Config, validateSchema } from './schema'

export const opencollective = (app: probot.Application): void => {
  app.log('OpenCollective Bot up!')

  app.on('issues.closed', async (context: probot.Context) => {
    console.log('HANDLING 1')

    /* Get config */

    const configuration = await getConfig<Config>(context, 'opencollective.yml')
    const validConfiguration = await validateSchema(configuration)

    if (!validConfiguration) {
      context.log.error("Coudln't load your configuration.")
      return
    }

    console.log('HANDLING 2', configuration)

    console.log('HANDLING 3', context.payload.issue.labels)

    /* Check labels */

    if (
      !!validConfiguration.labels &&
      context.payload.issue.labels.every(
        (label: { name: string }) =>
          !validConfiguration.labels!.includes(label.name),
      )
    ) {
      context.log.info(`OpenCollective Bot not configured for recieved labels.`)
      return
    }
    /* Post */

    console.log('HANDLING 4', context.payload.issues.labels)

    /* prettier-ignore */
    const body = `${validConfiguration.message}\n\n${validConfiguration.opencollective}`

    const issue = context.issue()

    console.log('HANDLING 5', issue)

    await context.github.issues.createComment({
      repo: issue.repo,
      owner: issue.owner,
      number: issue.number,
      body: body,
    })

    console.log('HANDLING 6 commented')

    context.log.info(`Commented on ${issue.owner}/${issue.repo}.`)
  })
}
