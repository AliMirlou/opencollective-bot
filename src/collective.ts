import fetch from 'isomorphic-fetch'

export type Tier = string

export type Member = {
  MemberId: number
  type: 'COLLECTIVE' | 'EVENT' | 'ORGANIZATION' | 'USER'
  role:
    | 'HOST'
    | 'ADMIN'
    | 'BACKER'
    | 'FOLLOWER'
    | 'MEMBER'
    | 'FUNDRAISER'
    | 'CONTRIBUTOR'
  tier?: Tier
  isActive: boolean
  name: string
  github: string | null
}

/**
 *
 * Fetches members of a particuar collective.
 *
 * @param slug
 */
export async function getCollectiveMembers(slug: string): Promise<Member[]> {
  return fetch(`https://opencollective.com/${slug}/members.json`).then(res =>
    res.json(),
  )
}
