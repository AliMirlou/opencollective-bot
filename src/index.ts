import { createProbot, ApplicationFunction } from 'probot'
import logRequestErrors from 'probot/lib/middleware/log-request-errors'

import { opencollective } from './bot'

/* Credentials */

if (
  !process.env.APP_ID ||
  !process.env.WEBHOOK_SECRET ||
  !process.env.PRIVATE_KEY
) {
  throw new Error('Missing credentials.')
}

/* Probot setup */

const probot = createProbot({
  id: parseInt(process.env.APP_ID, 10),
  secret: process.env.WEBHOOK_SECRET,
  cert: process.env.PRIVATE_KEY,
  port: 3000,
})

/* Load apps */

const apps: ApplicationFunction[] = [
  opencollective,
  require('probot/lib/apps/default'),
  require('probot/lib/apps/sentry'),
  require('probot/lib/apps/stats'),
]

process.on('unhandledRejection', probot.errorHandler)

apps.forEach(appFn => probot.load(appFn))

// Register error handler as the last middleware
probot.server.use(logRequestErrors)

/* Start the server */

probot.start()
