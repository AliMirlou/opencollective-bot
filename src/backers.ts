import { uniq } from 'lodash'

import { Tier, Member } from './collective'
import { is } from './utils'

/**
 *
 * Returns tiers of a backer.
 *
 * @param name
 */
export function getIssueAuthorCollectiveTiers(
  collectiveMembers: Member[],
  issueAuthorGithubHandle: string,
  issueAuthorGithubOrganisations: string[],
): Tier[] | null {
  const tiers = collectiveMembers.reduce<Tier[] | null>((acc, member) => {
    if (!member.account.githubHandle || acc === null) return acc

    const isMember = [
      issueAuthorGithubHandle,
      ...issueAuthorGithubOrganisations,
    ].some(is(member.account.githubHandle))

    /* Tiers */
    if (isMember && member.role === 'BACKER' && member.tier) {
      return uniq([...acc, member.tier])
    }

    /* Admins, Contributors */
    if (isMember && ['HOST', 'ADMIN', 'CONTRIBUTOR'].includes(member.role)) {
      return null
    }

    return acc
  }, [])

  return tiers
}
