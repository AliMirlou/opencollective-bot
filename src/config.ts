import * as joi from 'joi'
import mls from 'multilines'
import * as probot from 'probot'

import { Tier } from './collective'
import { GithubLabel } from './github'
import { intersect } from './utils'

export type Config = {
  collective: string
  tiers: TierConfig[]
  invitation: Message
}

export type TierConfig = {
  tiers: Tier[] | '*'
  labels: GithubLabel[]
  message: Message
}

export type Message = string

/* Schema */

const tierConfigSchema = joi.object().keys({
  tiers: joi
    .array()
    .items(joi.string())
    .allow('*'),
  labels: joi.array().items(joi.string()),
  message: joi.string().required(),
})

export const configSchema = joi
  .object()
  .keys({
    collective: joi
      .string()
      .regex(/^[\w\-]+$/, 'Use your collective slug.')
      .required(),
    tiers: joi
      .array()
      .items(tierConfigSchema)
      .optional()
      .default([
        {
          tiers: '*',
          labels: ['priority'],
          message: mls`
          | Hey :wave:,
          | 
          | Thanks for backing our project. We will handle your issue
          | with priority support. To make sure we don't forget how special
          | you are, we added a ${'`priority`'} label to your issue.
          |
          | Thanks again for backing us :tada:!
          `,
        },
      ]),
    invitation: joi
      .string()
      .default(
        mls`
        | Hey :wave:,
        |
        | Thank you for opening an issue. We will get back to you as
        | soon as we can. Also, check out our OpenCollective and consider
        | backing us.
        |
        | <link>
        |
        | PS.: We offer ${'`priority`'} support for all backers. Don't forget to
        | add ${'`priority`'} label when you start backing us :smile:
      `,
      )
      .optional()
      .description('An invitation message shown to non-backers'),
  })
  .requiredKeys(['collective'])

/**
 *
 * Loads configuration from Github.
 *
 * @param context
 */
export async function getConfig(
  context: probot.Context,
): Promise<Config | null> {
  try {
    const config = await context.config('opencollective.yml')
    const value = await joi.validate(config, configSchema)

    return value
  } catch (err) {
    return null
  }
}

/**
 *
 * Returns a collection of all labels in configuration.
 *
 * @param config
 */
export function getLabelsFromConfig(config: Config): GithubLabel[] {
  return config.tiers.reduce<GithubLabel[]>(
    (acc, tier) => [...acc, ...tier.labels],
    [],
  )
}

/**
 *
 * Returns a collection of labels in configuration for
 * particular tiers..
 *
 * @param config
 * @param tiers
 */
export function getLabelsFromConfigForTiers(
  config: Config,
  tiers: Tier[],
): GithubLabel[] {
  return config.tiers.reduce<GithubLabel[]>((acc, tier) => {
    if (tier.tiers === '*') {
      return [...acc, ...tier.labels]
    } else if (intersect(tier.tiers)(tiers)) {
      return [...acc, ...tier.labels]
    } else {
      return acc
    }
  }, [])
}

/**
 *
 * Returns a collection of messages assignable to partiular tier.
 *
 * @param config
 * @param tiers
 */
export function getMessagesFromConfigForTiers(
  config: Config,
  tiers: Tier[],
): Message[] {
  const rawMessages = config.tiers.reduce<Message[]>((acc, tier) => {
    if (tier.tiers === '*') {
      return [...acc, tier.message]
    } else if (intersect(tier.tiers)(tiers)) {
      return [...acc, tier.message]
    } else {
      return acc
    }
  }, [])

  const messages = rawMessages.map(hydrateMessage)
  return messages

  /* Helper functions */
  /**
   *
   * Hydartes raw message to include placeholder values.
   *
   * @param message
   */
  function hydrateMessage(message: string): string {
    return message.replace(
      '<link>',
      `https://opencollective.com/${config.collective}`,
    )
  }
}
