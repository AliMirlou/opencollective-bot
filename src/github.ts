import Octokit, {
  Response,
  IssuesCreateCommentResponse,
  IssuesAddLabelsResponseItem,
} from '@octokit/rest'
import { Message } from './config'

export type GithubLabel = string

export interface GithubIssue {
  number: number
  owner: string
  repo: string
}

/**
 *
 * Returns a list of user organisations.
 *
 * @param context
 */
export async function getUserOrganisations(
  github: Octokit,
  username: string,
): Promise<string[]> {
  return github.orgs
    .listForUser({
      username: username,
    })
    .then(res => res.data.map(org => org.login))
}

/**
 *
 * Adds messages to a GitHub Issue.
 *
 * @param github
 * @param issue
 * @param messages
 */
export async function messageGithubIssue(
  github: Octokit,
  issue: GithubIssue,
  messages: Message[],
): Promise<Response<IssuesCreateCommentResponse>[]> {
  const actions = messages.map(message =>
    github.issues.createComment({
      repo: issue.repo,
      owner: issue.owner,
      issue_number: issue.number,
      body: message,
    }),
  )

  return Promise.all(actions)
}

/**
 *
 * Adds labels to a GitHub Issue and ensures they all exist.
 *
 * @param github
 * @param issue
 * @param messages
 */
export async function labelGithubIssue(
  github: Octokit,
  issue: GithubIssue,
  labels: GithubLabel[],
): Promise<Response<IssuesAddLabelsResponseItem[]>> {
  /**
   * Flow
   * 1. Check if all labels exist.
   * 2. Create missing ones.
   * 3. Add labels to issue.
   */

  const actions = labels.map(async label =>
    github.issues
      .getLabel({ ...issue, name: label })
      .then(() => Promise.resolve())
      .catch(() =>
        github.issues.createLabel({
          ...issue,
          name: label,
          color: 'fbca04',
        }),
      ),
  )

  await Promise.all(actions)

  return github.issues.addLabels({
    repo: issue.repo,
    owner: issue.owner,
    issue_number: issue.number,
    labels: labels,
  })
}

/**
 *
 * Removes labels from a GitHub Issue.
 *
 * @param github
 * @param issue
 * @param messages
 */
export async function removeLabelsFromGithubIssue(
  github: Octokit,
  issue: GithubIssue,
  labels: GithubLabel[],
): Promise<GithubLabel[]> {
  const actions = labels.map(label =>
    github.issues.removeLabel({
      repo: issue.repo,
      owner: issue.owner,
      issue_number: issue.number,
      name: label,
    }),
  )

  await Promise.all(actions)
  return labels
}
