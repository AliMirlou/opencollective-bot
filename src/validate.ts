import bodyParser from 'body-parser'
import express from 'express'
import { Server } from 'http'
import * as joi from 'joi'
import yaml from 'js-yaml'
import mls from 'multilines'
import { AddressInfo } from 'net'

import { configSchema } from './config'

/* Server */

const server = express()

server.get('*', (req, res) => {
  res.send(mls`
  | <pre>Usage:
  |
  | curl -X POST --data-binary @opencollective.yml https://opencollective-bot.now.sh/validate
  | <pre>
  `)
})

server.post(
  '*',
  bodyParser.urlencoded({ extended: false }),
  async (req, res) => {
    try {
      // Load file.
      const file = Object.keys(req.body)[0]
      const config = yaml.safeLoad(file)

      // Validate configuration.
      joi
        .validate(config, configSchema, {
          abortEarly: false,
        })
        .then(() => {
          res.status(200).send('Valid configuration!')
        })
        .catch(err => {
          /* Compose a report */
          const message = err.details
            .map(
              (detail: { path: string[]; message: string }) =>
                `- ${detail.path.join('.')}: ${detail.message}`,
            )
            .join('\n')

          res.status(400).send(message)
        })
    } catch (err) {
      res.status(400).send('Something went wrong.')
    }
  },
)

/* Start */

/* istanbul ignore if */
if (process.env.NODE_ENV !== 'test') {
  const http = main(4000)
  const port = (http.address() as AddressInfo).port

  console.log(`Server running on http://localhost:${port}`)
}

/**
 * Starts the server
 */
export function main(port: number): Server {
  return server.listen(port)
}
