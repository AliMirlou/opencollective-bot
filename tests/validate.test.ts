import { safeDump } from 'js-yaml'
import { AddressInfo, Server } from 'net'
import request from 'request-promise-native'

import { main } from '../src/validate'

describe('validate', () => {
  let uri: string
  let server: Server

  beforeAll(() => {
    server = main(0)
    const { port } = server.address() as AddressInfo

    uri = `http://localhost:${port}/`
  })

  afterAll(() => {
    server.close()
  })
  test('shows correct usage description', async () => {
    const res = await request({
      uri,
      method: 'GET',
      json: true,
    }).promise()

    expect(res).toMatchSnapshot()
  })

  test('correctly reports invalid configuration', async () => {
    const config = safeDump({
      opencollective: 'https://opencollective.com/graphql-shield',
    })

    await request({
      uri,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: config,
    })
      .promise()
      .then(() => {
        fail()
      })
      .catch(err => {
        expect(err).toMatchSnapshot()
      })
  })

  test('correctly handles invalid input', async () => {
    const config = `f123
      opencollective: 'https://opencollective.com/graphql-shield',
    `

    await request({
      uri,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: config,
    })
      .promise()
      .then(() => {
        fail()
      })
      .catch(err => {
        expect(err).toMatchSnapshot()
      })
  })

  test('correctly validates valid configuration', async () => {
    const config = safeDump({
      collective: 'graphql-shield',
    })

    const res = await request({
      uri,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: config,
    })

    expect(res).toMatchSnapshot()
  })
})
