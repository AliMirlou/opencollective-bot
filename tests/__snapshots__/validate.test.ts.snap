// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate correctly handles invalid input 1`] = `[StatusCodeError: 400 - "Something went wrong."]`;

exports[`validate correctly reports invalid configuration 1`] = `[StatusCodeError: 400 - "- collective: \\"collective\\" is required\\n- opencollective: \\"opencollective\\" is not allowed"]`;

exports[`validate correctly validates valid configuration 1`] = `"Valid configuration!"`;

exports[`validate shows correct usage description 1`] = `
"<pre>Usage:

curl -X POST --data-binary @opencollective.yml https://opencollective-bot.now.sh/validate
<pre>"
`;
